Add basic support for SigMF input

kerel-fs opened the pull request:
Follows the proposal from #720.

Currently supported & tested data types:
- `ci16_le`

This list can probably be extended quite easily, for the first iteration I wanted to focus on what I can test.

daniestevez commented on apps/gr_satellites:
```diff
@@ -292,6 +297,61 @@ class gr_satellites_top_block(gr.top_block):
+                file=sys.stderr)
+            sys.exit(1)
+        if not self.options.sigmf_in.endswith('sigmf-meta'):
+            print("SIGMF_IN has invalid file extension, expected '.sigmf-meta'", file=sys.stderr)
```
Maybe we can print the actual `self.options.sigmf_in` that the user passed instead of the placeholder `SIGMF_IN`. That would make it even more obvious for the user what is wrong with the CLI arguments.

daniestevez commented on apps/gr_satellites:
```diff
@@ -292,6 +297,61 @@ class gr_satellites_top_block(gr.top_block):
+    def setup_sigmf_input(self):
+        if self.options.sigmf_in.endswith('sigmf'):
+            print(
+                "Failed to load SigMF archive (archives are not yet supported). "
```
There is an inconsistent usage of double quotes (`"`) and simple quotes (`'`). In this PR. I believe that the gr-satellites code base only uses simple quotes to denote strings.

daniestevez commented on apps/gr_satellites:
```diff
@@ -292,6 +297,61 @@ class gr_satellites_top_block(gr.top_block):
+                sys.exit(1)
+
+        datatype = meta['global']['core:datatype']
+        dataset_filename = self.options.sigmf_in[:-11] + '.sigmf-data'
```
Maybe this, as it makes it more obvious where the number `11` comes from:
```suggestion
        dataset_filename = self.options.sigmf_in[:-len('.sigmf-meta')] + '.sigmf-data'
```

daniestevez commented on apps/gr_satellites:
```diff
@@ -292,6 +297,61 @@ class gr_satellites_top_block(gr.top_block):
+            print(f"Unsupported SigMF Dataset format of stored samples ('{datatype}'). Supported datatypes are: {', '.join(datatypes_supported)}.", file=sys.stderr)
+            sys.exit(1)
+
+        if 'captures' in meta and 'core:datetime' in meta['captures']:
+            self.options.start_time = meta['captures']['core_datetime']
+
+        if 'capture' in meta and 'core:sample_start' in meta['captures'] and meta['captures']['core:sample_start'] != 0:
+            print('SigMF Capture Segments with non-zero sample start index are not supported (yet).', file=sys.stderr)
+            sys.exit(1)
```
Some comments about captures.

1.  `if 'capture' in meta` and `'core_datetime'` look like typos.

2. [According to the SigMF spec](https://sigmf.org/#sigmf-metadata-format), "The top-level Object MUST contain three JSON Objects named `global`, `captures`, and `annotations`.", so we should fail and exit if `'captures' not in meta`.

3. `meta['captures']` is going to be a `list` (even if there is a single capture). I would probably only set `self.options.start_time` if there is a single capture that contains `core:datetime` and has `core:sample_start` equal to zero. I would allow other types of captures (as long as they are not malformed SigMF metadata), but not set the `self.options.start_time` in that case (and perhaps, if there was a `core:datetime` somewhere, print a warning that this `datetime` was not applied). Being tolerant in what we accept might be useful, if for instance a user has a dataset that they want to process, but the dataset contains multiple captures with time discontinuities. The user might not care about setting `start_time`, so we should allow them to process this dataset.

4. [According to the SigMF spec](https://sigmf.org/#subsec:CapturesArray), `core:sample_start` is a required field in each capture, so probably we want to fail here if `core:sample_start` is not present.

kerel-fs replied:
1. typo fixed (good catch, apparently I never tested if the `start_time` is properly applied).
2. My original idea was to enforce only what is strictly necessary for gr-satellites, and be permissive regarding non-compliant metadata files when possible. However, by enforcing SigMF spec compliance the parser does become simpler. So, I've added a check to enforce `global`, `captures` and `annotations` as top-level objects. 
3. implemented as you described.
4. Done. Note that `[]` is also a valid value for the `captures` object. So either `captures` is `[]`, or we require `core:sample_start`.

kerel-fs commented:
Updated the PR addressing your comments, including the suggestions 1 & 2 from https://github.com/daniestevez/gr-satellites/issues/720#issuecomment-2673987186.

daniestevez reviewed (approved):
This looks great now. Approved and merging.
